hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -122,18 +121,14 @@ public static List<FileSlice> getLatestFileSlicesForPartition(
    * @param hoodieTable instance of {@link HoodieTable} of interest
    * @return the list of Pairs of partition path and fileId
    */
-  public static List<Pair<String, HoodieBaseFile>> getLatestBaseFilesForAllPartitions(final List<String> partitions,
-                                                                                      final HoodieEngineContext context,
-                                                                                      final HoodieTable hoodieTable) {
+  public static HoodieData<Pair<String, HoodieBaseFile>> getLatestBaseFilesForAllPartitions(final HoodieData<String> partitions,
```

Comment:
Previously, we were collecting the distinct partitions list on the driver and then creating a new dataset with these partitions that we then fetch the base files for. Now I've left this to all execute on the executors and only collect when we need an aggregated result for the range pruning.

Reply:
+1

Reply:
If I am not wrong, the returned HoodieData here contains the same number of spark partitions as `HoodieData<String> partitions` which is likely number of hudi partitions. but we return base files from this method. and prior to this patch, the caller or somewhere down the line, we might parallelize based on number of base files (since we would be operating on list). But now, I guess we are resorting to use numSparkPartitions from HoodieData<Pair<String, HoodieBaseFile>>. 

So, for partitions having large no of base files, that might have perf impact. Can we ensure we use bloom index shuffle parallelism or inputRecords.numPartitions (passed in as an arg) so that we adaptively choose a value relative to incoming records